### PR TITLE
Publish snowplow events on CSV append success / failure

### LINKED ISSUE #39867: Publish snowplow events when appending to uploads
See the corresponding event that we are already sending when uploads are created.

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes https://github.com/metabase/metabase/issues/39867

### Description

This takes "heavy inspiration" from the initial upload events to give analytics coverage for appends as well - the only salient difference is that we don't refer back to the original model created when the table was first created.

### How to verify

... I'd actually appreciate a pointer here, I'm not sure how to test all the wiring. 

Could even have missed something, in fact.

### CODE REVIEW

File: snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0
Comment (qnkhuat) [inline snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0:1]: I'm not sure if it's acceptable to modify the enum of a snowplow event, but generally, you can't change a version once it's published; you're supposed to create a new version for it.

so you either can confirm of you can modify enum of a version in #external-snowcat-cloud, or you can do it the normal way by creating a new schema

File: test/metabase/upload_test.clj
Comment (qnkhuat) [inline test/metabase/upload_test.clj:1523]:
```suggestion
         (is (=? {:data {"size_mb"         (mt/malli=? pos-int?)
```
I'd avoid testing the exact size here

**Review by qnkhuat — CHANGES_REQUESTED** [review]
I think you'll need to create a new schema for the added enum value and we need to assoc `model_id` to the snowplow event data

File: snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0
Comment (crisptrutski) [reply snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0:1]: I guess I just assumed we could add values - will follow up and confirm. 

Is there a way I can test out these events in practice before merging?

File: test/metabase/upload_test.clj
Comment (crisptrutski) [reply test/metabase/upload_test.clj:1523]: Fair enough, there could be some platform sensitivity. What do you think about matching the order of magnitude eat least?

File: src/metabase/upload.clj
Comment (crisptrutski) [reply src/metabase/upload.clj:705]: We don't have a model in scope here - a model is created in `create-csv-upload!` when we first create the table, but the append action runs directly against the table. 

We could search for the first model created against this table, but that would be brittle as the model could have been archived, deleted, or edit in some semantic way.

File: snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0
Comment (qnkhuat) [reply snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-0:1]: Try this https://www.notion.so/metabase/Snowplow-integration-5da1f874beda4153b4fccfa6c1e77caa?pvs=4#6665b4c007d64d1d8e45c5d4c423ac29

**Comment by crisptrutski** [comment]
* **#39951** <a href="https://app.graphite.dev/github/pr/metabase/metabase/39951?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#39944** <a href="https://app.graphite.dev/github/pr/metabase/metabase/39944?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @crisptrutski and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: test/metabase/upload_test.clj
Comment (crisptrutski) [reply test/metabase/upload_test.clj]: I've decided to leave this using the exact size:

1. Our initial upload case uses an explicit number.
2. We don't really have any other tests of the size calculation helper.
3. It seems to be deterministic at least on unix platforms.
4. It's easy to update if the value changes.

File: src/metabase/analytics/snowplow.clj
Comment (crisptrutski) [inline src/metabase/analytics/snowplow.clj:35]: According to the process doc, we wait until the commit is merged and only then engage our Snowcat Cloud partners to update it. 

So AFAIK this branch is g2g

File: snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-1
Comment (qnkhuat) [inline snowplow/iglu-client-embedded/schemas/com.metabase/csvupload/jsonschema/1-0-1:8]:
```suggestion
    "version": "1-0-1"
```

File: src/metabase/upload.clj
Comment (qnkhuat) [reply src/metabase/upload.clj:702]: hmm, I think the model_id is desirable. I'd see a use case for it in which we want to know how of  X uploads were created, how many of them are then appended.

Comment (crisptrutski) [reply src/metabase/upload.clj:702]: I'm not fond of guessing the ID, I don't think there's enough value. I'm not sure how we query the snowcat cloud database, but imagine we can join on the upload event there and it'd be more robust.

File: src/metabase/upload.clj
Comment (calherries) [inline src/metabase/upload.clj:696]: If we want to track the model_id that was uploaded from, that is only available on the frontend.

Comment (crisptrutski) [reply src/metabase/upload.clj:696]: Feel like I'm missing something in your comment, so apologies if my response doesn't quite speak to it.

We use the underlying table itself as the anchor resource for the event, not the model - this is consistent with the event published on initial upload. For the initial upload the `model_id` for the model that we create is put inside the `:details` map instead.

Using the Model whose page we append from on the frontend is not something I thought of, slightly different but compatible semantics. It still leaves us without a natural id to include if the action was taken from the Collections screen though, so I don't see the harm in just excluding it from both for now.

Comment (calherries) [reply src/metabase/upload.clj:698]: resolved on zoom

File: src/metabase/upload.clj
Comment (calherries) [reply src/metabase/upload.clj:704]: @qnkhuat 

> I'd see a use case for it in which we want to know how of X uploads were created, how many of them are then get appended.

We can calculate that statistic without the model_id, because there is a one-to-one relationship between models created from uploads and tables created from uploads.

We would know how many uploads were created: the number of model IDs from successful initial upload events.
We would know how many uploads have been appended to: the number of distinct table IDs from successful append events.

File: test/metabase/upload_test.clj
Comment (calherries) [inline test/metabase/upload_test.clj:1544]:
```suggestion
         (is (= {:data {"size_mb"     5.245208740234375E-5
```

File: test/metabase/upload_test.clj
Comment (calherries) [inline test/metabase/upload_test.clj:1519]: Can you leave some `testing` blocks here? One for the success case, and one for the failure case

**Review by calherries — APPROVED** [review]
LGTM

File: test/metabase/upload_test.clj
Comment (crisptrutski) [reply test/metabase/upload_test.clj:1519]: Good idea. I'll add for both creates and appends.

**Review by qnkhuat — APPROVED** [review]
okay I think it's okay to leave out model_id then.
LGTM now!

**Pull request merged**
